partitions.go

```diff
@@ -0,0 +1,232 @@
+}
+
+func (p *PartitionField) UnmarshalJSON(b []byte) error {
+	type Alias PartitionField
+	aux := struct {
+		TransformString string `json:"transform"`
+		*Alias
+	}{
+		Alias: (*Alias)(p),
+	}
+
+	err := json.Unmarshal(b, &aux)
+	if err != nil {
+		return err
+	}
+
+	if p.Transform, err = ParseTransform(aux.TransformString); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// PartitionSpec captures the transformation from table data to partition values
+type PartitionSpec struct {
+	// any change to a PartitionSpec will produce a new spec id
+	id     int
+	fields []PartitionField
+
+	// this is populated by initialize after creation
+	sourceIdToFields map[int][]PartitionField
+}
+
+func NewPartitionSpec(fields ...PartitionField) PartitionSpec {
+	return NewPartitionSpecID(InitialPartitionSpecID, fields...)
+}
+
+func NewPartitionSpecID(id int, fields ...PartitionField) PartitionSpec {
+	ret := PartitionSpec{id: id, fields: fields}
+	ret.initialize()
+	return ret
+}
+
+// CompatibleWith returns true if this partition spec is considered
+// compatible with the passed in partition spec. This means that the two
+// specs have equivalent field lists regardless of the spec id.
+func (ps *PartitionSpec) CompatibleWith(other *PartitionSpec) bool {
+	if ps == other {
+		return true
+	}
+
+	if len(ps.fields) != len(other.fields) {
+		return false
+	}
+
+	return slices.EqualFunc(ps.fields, other.fields, func(left, right PartitionField) bool {
+		return left.SourceID == right.SourceID && left.Name == right.Name &&
+			left.Transform == right.Transform
+	})
+}
+
+// Equals returns true iff the field lists are the same AND the spec id
+// is the same between this partition spec and the provided one.
+func (ps *PartitionSpec) Equals(other PartitionSpec) bool {
+	return ps.id == other.id && slices.Equal(ps.fields, other.fields)
+}
+
+func (ps PartitionSpec) MarshalJSON() ([]byte, error) {
+	if ps.fields == nil {
+		ps.fields = []PartitionField{}
+	}
+	return json.Marshal(struct {
+		ID     int              `json:"spec-id"`
+		Fields []PartitionField `json:"fields"`
+	}{ps.id, ps.fields})
+}
+
+func (ps *PartitionSpec) UnmarshalJSON(b []byte) error {
+	aux := struct {
+		ID     int              `json:"spec-id"`
+		Fields []PartitionField `json:"fields"`
+	}{ID: ps.id, Fields: ps.fields}
+
+	if err := json.Unmarshal(b, &aux); err != nil {
+		return err
+	}
+
+	ps.id, ps.fields = aux.ID, aux.Fields
+	ps.initialize()
+	return nil
+}
+
+func (ps *PartitionSpec) initialize() {
+	ps.sourceIdToFields = make(map[int][]PartitionField)
+	for _, f := range ps.fields {
+		ps.sourceIdToFields[f.SourceID] =
+			append(ps.sourceIdToFields[f.SourceID], f)
+	}
+}
+
+func (ps *PartitionSpec) ID() int                    { return ps.id }
+func (ps *PartitionSpec) NumFields() int             { return len(ps.fields) }
+func (ps *PartitionSpec) Field(i int) PartitionField { return ps.fields[i] }
+
+func (ps *PartitionSpec) IsUnpartitioned() bool {
```

Comment:
just to make sure: this will return true only if 1) fields are empty or 2) all transforms are void

Reply:
Correct, I've added tests accordingly.